### ISSUE TITLE
Some questions about reproducing the origin paper experiments

### ISSUE THREAD

**Issue by challow0** [michaelrzhang/lookahead#5]
Hi, 
     I am a student who is very interested in your lookahead work, so currently I am reproducing your paperwork. Trying to identify the performance of the lookahead optimizer. Now I am doing this  `https://github.com/salesforce/awd-lstm-lm`  experiment. And I encounter some questions and really need you to help me to solve it.
     **Problem 1:**
     Following is the changed code, I don't know the **valid_loss** and the **train_loss** I chose is correct or not! Can you help me to know the **training loss** and **valid loss** you used? 

```python
def evaluate(data_source, batch_size=10,test=False):
    # Turn on evaluation mode which disables dropout.
    model.eval()
    if args.model == 'QRNN': model.reset()
    epoch_loss = 0 # running mean
    total_loss = 0 # running sum
    hidden = model.init_hidden(batch_size)
    batch = 0
    for i in range(0, data_source.size(0) - 1, args.bptt): #batch loop
        data, targets = get_batch(data_source, i, args, evaluation=True)
        output, hidden = model(data, hidden)
        total_loss += len(data) * criterion(model.decoder.weight, model.decoder.bias, output, targets).data
        hidden = repackage_hidden(hidden)

        # -------Different 1 ----------
        if test == False:
            epoch_loss = (epoch_loss*batch + (criterion(model.decoder.weight, model.decoder.bias, output, targets).data)).item() / (batch+1) #            
            batch += 1
    if test == False:
        global valid_loss, valid_ppl, valid_bpc
        valid_loss = np.append(valid_loss, epoch_loss)
        valid_ppl = np.append(valid_ppl, np.exp(epoch_loss))
        valid_bpc = np.append(valid_bpc, epoch_loss / np.log(2))
    return total_loss.item() / len(data_source)


def train():
    # Turn on training mode which enables dropout.
    if args.model == 'QRNN': model.reset()
    epoch_loss = 0 # running mean (only for drawing the curve)
    total_loss = 0 # running sum
    start_time = time.time()
    hidden = model.init_hidden(args.batch_size)
    batch, i = 0, 0
    while i < train_data.size(0) - 1 - 1: #batch loop
        bptt = args.bptt if np.random.random() < 0.95 else args.bptt / 2.
        # Prevent excessively small or negative sequence lengths
        seq_len = max(5, int(np.random.normal(bptt, 5)))
        # There's a very small chance that it could select a very long sequence length resulting in OOM
        # seq_len = min(seq_len, args.bptt + 10)

        lr2 = optimizer.param_groups[0]['lr']
        optimizer.param_groups[0]['lr'] = lr2 * seq_len / args.bptt
        model.train()
        data, targets = get_batch(train_data, i, args, seq_len=seq_len)

        # Starting each batch, we detach the hidden state from how it was previously produced.
        # If we didn't, the model would try backpropagating all the way to start of the dataset.
        hidden = repackage_hidden(hidden)
        optimizer.zero_grad()

        output, hidden, rnn_hs, dropped_rnn_hs = model(data, hidden, return_h=True)
        raw_loss = criterion(model.decoder.weight, model.decoder.bias, output, targets)

        loss = raw_loss
        # Activiation Regularization
        if args.alpha: loss = loss + sum(args.alpha * dropped_rnn_h.pow(2).mean() for dropped_rnn_h in dropped_rnn_hs[-1:]) #alpha=2
        # Temporal Activation Regularization (slowness)
        if args.beta: loss = loss + sum(args.beta * (rnn_h[1:] - rnn_h[:-1]).pow(2).mean() for rnn_h in rnn_hs[-1:]) #beta=1
        loss.backward()

        # `clip_grad_norm` helps prevent the exploding gradient problem in RNNs / LSTMs.
        if args.clip: torch.nn.utils.clip_grad_norm_(params, args.clip) #clip=0.25
        optimizer.step()

        total_loss += raw_loss.data # running sum loss
        optimizer.param_groups[0]['lr'] = lr2 # retrieves the lr after it was changed de pending on the sequence length
        if batch % args.log_interval == 0 and batch > 0 and args.optimizer == 'nt_asgd': #
            cur_loss = total_loss.item() / args.log_interval
            elapsed = time.time() - start_time
            print('| epoch {:3d} | {:5d}/{:5d} batches | lr {:05.5f} | ms/batch {:5.2f} | '
                    'loss {:5.2f} | ppl {:8.2f} | bpc {:8.3f}'.format(
                epoch, batch, len(train_data) // args.bptt, optimizer.param_groups[0]['lr'],
                elapsed * 1000 / args.log_interval, cur_loss, math.exp(cur_loss), cur_loss / math.log(2)))
            total_loss = 0
            start_time = time.time()
        
        epoch_loss = (epoch_loss * batch + loss.item()) / (batch + 1) # for drawing purposes
        batch += 1
        i += seq_len
    # ------Different 2--------
    global tr_loss, tr_ppl, tr_bpc
    # Version 1
    tr_loss = np.append(tr_loss, epoch_loss)
    tr_ppl = np.append(tr_ppl, np.exp(epoch_loss))
    tr_bpc = np.append(tr_bpc, epoch_loss / np.log(2))
    # Version 2
    tr_loss = np.append(tr_loss, loss.item())
    tr_ppl = np.append(tr_ppl, np.exp(loss.item()))
    tr_bpc = np.append(tr_bpc, loss.item() / np.log(2))
```

**Problem 2:**
    I just running several experiments, and find the following result ~ The train_ppl is calculated from **the Upper Version 1 train_loss**. I don't know which thing happened to cause the lookahead optimizer performs not very good. If you can give me some advice to improve the lookahead optimizer performance? Thank you very much!
     

OPtimizer | lr |  la_alpha | la_steps  | Train | Val | Test
-- | -- | -- | -- | -- | -- | --
SGD | 30 | - | - | 49.06 | 66.43 | 63.41
LA(SGD) | 30 | 0.8 | 10 | 53.82 | 67.82 | 64.62
ADAM | 0.005 | - | - | 39.79 | 63.52 | 60.41
LA(Adam) | 0.005 | 0.8 | 10 | 41.64 | 63.69 | 60.71

**Comment by AtheMathmo** [comment]
Hi @challow0 

Thanks for your post.

With regards to the first problem, when reporting the loss we evaluate the model over an entire epoch of both the training, validation, and test data after training has concluded. In the code snipper you included, the training loss is computed on-the-fly as the model weights are update. However, I'd expect the `epoch_loss` that computes a running average of the loss would be close enough later in training.

For performance, did you follow the procedure described in Appendix C.3 of our paper? We first tuned the base optimizer and then tuned lookahead on top (trying only a small handful of settings for LA). Looking at your table, the base optimizers seem to be _about_ right (a little worse than we found), but LA is performing quite a bit worse than we expect given the baseline.

I can try to dig up the optimal settings that we found for each optimizer in these experiments. If indeed you have followed our tuning procedure, we might be able to compare the final optimizer settings that we each used...

**Comment by challow0** [comment]
@AtheMathmo Thank you for your reply.

About the first problem, could you please show me your source codes, I am afraid I may code the mistaken codes again?

About the second problem, I just followed the procedure described in Appendix C.3 of the paper. Firstly try to find the best base optimizer parameter settings and then using these parameters in LA optimizer experiments. And find the best `la_steps and la_alpha`.

The following is my settings, 

 3 **(nlayers)** LSTM layers of 1150 **(nhid)** hidden units each, embedding layer size equal to 400 **(emsize)**, gradient clipping of maximum norm 0.25 **(clip)**, batch size equal to 80 **(batch_size)**, weight decay of 1.2e-6 **(wdecay)**, L2 regularisation on RNN activations equal to 2 **(alpha)** and temporal activation regularisation equal to 1 **(beta)**. Dropout probability on the word vectors (input embedding layer) equal to 0.65 **(dropouti)**, dropout probability to remove words from the embedding layer equal to 0.1 **(dropoute)**, dropout probability on RNN layers (within each hidden layer) equal to 0.3 **(dropouth)** and dropout probability for the weight-drop method (applied to the RNN hidden to hidden matrix) equal to 0.5 **(wdrop)**. All the epochs using in the experiments are 500 **(epochs)**.

**This is the performance of SGD optimizer with different learning rate, showing the learning rate equals to 30 best**
![Different_Learning_Rate_SGD](https://user-images.githubusercontent.com/46772513/98339829-51ad2280-2047-11eb-8bc4-ad15566eac45.png)
.
**This is the performance of Adam optimizer with different learning rate, showing the learning rate equals to 0.005 best**
![Different_Learning_Rate_Adam](https://user-images.githubusercontent.com/46772513/98340175-eadc3900-2047-11eb-9b40-afbea7ed73b5.png)

Knowing the best parameters in base optimizers, next I tried the different  **la_alpha and la_steps** using in the LA(SGD) (lr=30) and the LA(Adam) (lr=0.005). 
`la_alpha=0.8 and la_steps=10 performce best in LA(Adam),  la_alpha=0.8 and la_steps=5 performce best in LA(SGD).`

But the performances just are not as well as yours, even comparing to the base optimizer, it performs worse...So here it is. Could you show me some advice?

**Comment by AtheMathmo** [comment]
Sorry for the lack of response. Things got quite busy on my end but I do intend to come back to this and take a careful look (+ share sourcecode)

**Comment by challow0** [comment]
@AtheMathmo Thanks for your time to do this thing ~ I am looking forward to your sharing. Wish you have a good day.

**Comment by densechen** [comment]
@challow0 Did you solve this problem?

**Comment by challow0** [comment]
> @challow0 Did you solve this problem?

No, I'm still waittttttting for the author's reply~~

**Comment by AtheMathmo** [comment]
Hi all,

Sorry for the (very) slow response. I spent some time trying to dig up the code we used to run these experiments (and the checkpoints). I believe that I have the correct code snapshot from the original repository, and have uploaded this into a new repository: https://github.com/AtheMathmo/lookahead-lstm . I haven't tested this code. There is some development code there, some unused things, etc.

In general, it has not been tidied up but I wanted to share something as I keep putting this off... The code also relies in some small ways on our larger development code base for handling utilities like logging and config management. I haven't included this but hopefully, you can get what you need from the code that is present.

I could not recover the checkpoints. Unfortunately, these were run on a google cloud instance and I no longer have access to this data.

In general, I'm still a bit perplexed by the difference in performance (for baseline models, and lookahead). With some luck, I'll have time this week to rerun this experiment with the code base I shared. Thanks for your patience!

**Comment by AtheMathmo** [comment]
Oh, and the code was run with pytorch 0.4. Let me know if you have any other questions --- I know the code is pretty ugly!

**Comment by challow0** [comment]
Hahaha, never mind about it. I know you may be busy with your works. Luckily getting your sharing, and I will try the code and figure out where is my problems. 
Anyway, thank you for your sharing. Have a nice day ~~
